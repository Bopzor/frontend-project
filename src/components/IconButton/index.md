Icon buttons are commonly found in app bars and toolbars.

Icons are also appropriate for toggle buttons that allow a single choice to be selected or deselected, such as
adding or removing a star to an item.

> Please import your icons from the
> [@gojob/icons (front-office-icons) library](/#404).

```js
const JustifyAll = require('../Icons/JustifyAll').default;
const Search = require('../Icons/Search').default;
const Coat = require('../Icons/Coat').default;
const User = require('../Icons/User').default;
const Graduate = require('../Icons/Graduate').default;
<div style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-around' }}>
  <IconButton>
    <JustifyAll />
  </IconButton>
  <IconButton>
    <Search />
  </IconButton>
  <IconButton>
    <Coat />
  </IconButton>
  <IconButton>
    <User />
  </IconButton>
  <IconButton>
    <Graduate />
  </IconButton>
</div>;
```

You may make the `<IconButton />` clickable by binding `onClick` property

```js
const JustifyAll = require('../Icons/JustifyAll').default;
<IconButton onClick={() => alert('That tickles!')}>
  <JustifyAll />
</IconButton>;
```

You may specify various sizes of `<IconButton />` by using `size` property.

```js
const JustifyAll = require('../Icons/JustifyAll').default;
const Search = require('../Icons/Search').default;
const Coat = require('../Icons/Coat').default;
const User = require('../Icons/User').default;
const Graduate = require('../Icons/Graduate').default;
<div style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-around' }}>
  <IconButton size="xs">
    <JustifyAll />
  </IconButton>
  <IconButton size="sm">
    <Search />
  </IconButton>
  <IconButton size="md">
    <Coat />
  </IconButton>
  <IconButton size="lg">
    <User />
  </IconButton>
  <IconButton size="xl">
    <Graduate />
  </IconButton>
</div>;
```

You can also change the background and foreground color of `<IconButton />` by specifying `backgroundColor`
and `color` props respectively.

```js
const JustifyAll = require('../Icons/JustifyAll').default;
const Search = require('../Icons/Search').default;
const Coat = require('../Icons/Coat').default;
const User = require('../Icons/User').default;
const Graduate = require('../Icons/Graduate').default;
<div style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-around' }}>
  <IconButton backgroundColor="alert" color="primary">
    <JustifyAll />
  </IconButton>
  <IconButton backgroundColor="success" color="disabled">
    <Search />
  </IconButton>
  <IconButton backgroundColor="primary" color="white">
    <Coat />
  </IconButton>
  <IconButton backgroundColor="transparent" color="melrose">
    <User />
  </IconButton>
  <IconButton backgroundColor="melrose" color="input">
    <Graduate />
  </IconButton>
</div>;
```
