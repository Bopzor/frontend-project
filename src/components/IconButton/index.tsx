import * as React from 'react';

import { hot } from 'react-hot-loader';

import { css, cx } from 'emotion';
import memoizeOne from 'memoize-one';
import { Omit } from 'react-redux';

import { Consumer, Theme } from '../../theme';

/**
 * Props for IconButton component
 */
export type IconButtonProps = {
  /**
   * Size of button
   */
  size?: keyof Theme['sizes'];
  /**
   * Background color
   */
  backgroundColor?: keyof Theme['styles']['bgGradients'];
  /**
   * Foreground color
   */
  color?: keyof Theme['colors'];
  /**
   * Ref to the underlying span (no onClick property) or button (with onClick property) component
   */
  innerRef?: React.Ref<HTMLSpanElement | HTMLButtonElement>;
} & (Omit<React.HTMLProps<HTMLButtonElement>, 'size'> | Omit<React.HTMLProps<HTMLSpanElement>, 'size'>);

/**
 * IconButton component is a circle shaped button for small content
 */
const IconButton = ({
  size = 'sm',
  backgroundColor = 'primary',
  color,
  className,
  onClick,
  innerRef,
  ...props
}: IconButtonProps) => {
  color = color || (backgroundColor === 'transparent' ? 'icon' : 'white');
  return (
    <Consumer>
      {(theme) => {
        const classNames = cx(styles(theme, size, backgroundColor, color, !!onClick), className);
        if (!onClick) {
          return <span ref={innerRef} className={classNames} {...props} />;
        }
        return <button ref={innerRef} className={classNames} onClick={onClick} {...props as any} />;
      }}
    </Consumer>
  );
};

export default hot(module)(IconButton);

const styles = memoizeOne(
  (
    { styles: { bgGradients }, colors, sizes }: Theme,
    size: IconButtonProps['size'],
    variant: IconButtonProps['backgroundColor'],
    color: IconButtonProps['color'],
    clickable?: boolean
  ) => {
    const length = sizes[size!];
    return css`
      display: flex;
      justify-content: center;
      align-items: center;
      border-radius: 50%;
      padding: 0;
      border: none;
      width: ${length}px;
      height: ${length}px;
      color: ${colors[color!]};
      cursor: ${clickable ? 'pointer' : 'default'};
      pointer-events: ${clickable ? 'all' : 'none'};
      ${bgGradients[variant!]};
      & > * {
        width: ${length - 4}px;
        height: ${length - 4}px;
      }
    `;
  }
);
