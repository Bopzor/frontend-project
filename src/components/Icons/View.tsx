import * as React from 'react'; // tslint:disable:max-line-length

const View = (props: React.SVGProps<SVGSVGElement>) => (
  <svg width={24} height={24} fill="currentColor" viewBox="0 0 32 32" {...props}>
    <path d="M16.003 24.01c-7.338 0-10.983-5.836-11.89-7.539a.974.974 0 0 1 .003-.927l-.003.005c.907-1.703 4.552-7.539 11.89-7.539s10.983 5.836 11.89 7.539a.974.974 0 0 1-.002.927l.003-.005c-.907 1.703-4.552 7.539-11.89 7.539zm.007-13.5a5.5 5.5 0 1 0 5.5 5.5 5.5 5.5 0 0 0-5.5-5.5zm0 9.001a3.5 3.5 0 1 1 3.5-3.5 3.5 3.5 0 0 1-3.5 3.5z" />
  </svg>
);

export default View;
