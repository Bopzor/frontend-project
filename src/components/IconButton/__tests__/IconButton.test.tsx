import * as React from 'react';

import { mount, render } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';

import IconButton from '../';
import View from '../../Icons/View';

describe('IconButton', () => {
  describe('Should match snapshot for', () => {
    test('default', () => {
      const cheerio = render(<IconButton />);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('clickable', () => {
      const cb = jest.fn();
      const cheerio = render(<IconButton onClick={cb} />);
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('with icon', () => {
      const cheerio = render(
        <IconButton>
          <View />
        </IconButton>
      );
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
    test('with colors', () => {
      const cheerio = render(
        <div>
          <IconButton backgroundColor="alert" color="primary">
            <View />
          </IconButton>
          <IconButton backgroundColor="success" color="disabled">
            <View />
          </IconButton>
          <IconButton backgroundColor="primary" color="white">
            <View />
          </IconButton>
          <IconButton backgroundColor="transparent" color="melrose">
            <View />
          </IconButton>
          <IconButton backgroundColor="melrose" color="input">
            <View />
          </IconButton>
        </div>
      );
      expect(renderToJson(cheerio)).toMatchSnapshot();
    });
  });
  test('should be clickable button with onClick prop', () => {
    const cb = jest.fn();
    const cheerio = mount(<IconButton onClick={cb} />);
    expect(cheerio.find('button')).toHaveLength(1);
    cheerio.simulate('click');
    expect(cb).toBeCalled();
  });
});
