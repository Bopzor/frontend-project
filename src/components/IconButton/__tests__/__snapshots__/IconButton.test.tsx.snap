// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton Should match snapshot for clickable 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  padding: 0;
  border: none;
  width: 24px;
  height: 24px;
  color: #ffffff;
  cursor: pointer;
  pointer-events: all;
  background: #3a44e9;
  background: linear-gradient(122deg,#3a44e9,#7038e3);
}

.emotion-0 > * {
  width: 20px;
  height: 20px;
}

<button
  class="emotion-0"
/>
`;

exports[`IconButton Should match snapshot for default 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  padding: 0;
  border: none;
  width: 24px;
  height: 24px;
  color: #ffffff;
  cursor: default;
  pointer-events: none;
  background: #3a44e9;
  background: linear-gradient(122deg,#3a44e9,#7038e3);
}

.emotion-0 > * {
  width: 20px;
  height: 20px;
}

<span
  class="emotion-0"
/>
`;

exports[`IconButton Should match snapshot for with colors 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  padding: 0;
  border: none;
  width: 24px;
  height: 24px;
  color: #ffffff;
  cursor: default;
  pointer-events: none;
  background: #3a44e9;
  background: linear-gradient(122deg,#3a44e9,#7038e3);
}

.emotion-2 > * {
  width: 20px;
  height: 20px;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  padding: 0;
  border: none;
  width: 24px;
  height: 24px;
  color: #5856d6;
  cursor: default;
  pointer-events: none;
  background: #ff418b;
  background: linear-gradient(to right,#ff418b,#ff3466);
}

.emotion-0 > * {
  width: 20px;
  height: 20px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  padding: 0;
  border: none;
  width: 24px;
  height: 24px;
  color: #ced1e2;
  cursor: default;
  pointer-events: none;
  background: #23bcba;
  background: linear-gradient(135deg,#23bcba,#45e994);
}

.emotion-1 > * {
  width: 20px;
  height: 20px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  padding: 0;
  border: none;
  width: 24px;
  height: 24px;
  color: #908eff;
  cursor: default;
  pointer-events: none;
  background: transparent;
}

.emotion-3 > * {
  width: 20px;
  height: 20px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  padding: 0;
  border: none;
  width: 24px;
  height: 24px;
  color: #f1f2f8;
  cursor: default;
  pointer-events: none;
  background: #9890e3;
  background: linear-gradient(315deg,#9890e3,#b1f4cf);
}

.emotion-4 > * {
  width: 20px;
  height: 20px;
}

<div>
  <span
    class="emotion-0"
  >
    <svg
      fill="currentColor"
      height="24"
      viewBox="0 0 32 32"
      width="24"
    >
      <path
        d="M16.003 24.01c-7.338 0-10.983-5.836-11.89-7.539a.974.974 0 0 1 .003-.927l-.003.005c.907-1.703 4.552-7.539 11.89-7.539s10.983 5.836 11.89 7.539a.974.974 0 0 1-.002.927l.003-.005c-.907 1.703-4.552 7.539-11.89 7.539zm.007-13.5a5.5 5.5 0 1 0 5.5 5.5 5.5 5.5 0 0 0-5.5-5.5zm0 9.001a3.5 3.5 0 1 1 3.5-3.5 3.5 3.5 0 0 1-3.5 3.5z"
      />
    </svg>
  </span>
  <span
    class="emotion-1"
  >
    <svg
      fill="currentColor"
      height="24"
      viewBox="0 0 32 32"
      width="24"
    >
      <path
        d="M16.003 24.01c-7.338 0-10.983-5.836-11.89-7.539a.974.974 0 0 1 .003-.927l-.003.005c.907-1.703 4.552-7.539 11.89-7.539s10.983 5.836 11.89 7.539a.974.974 0 0 1-.002.927l.003-.005c-.907 1.703-4.552 7.539-11.89 7.539zm.007-13.5a5.5 5.5 0 1 0 5.5 5.5 5.5 5.5 0 0 0-5.5-5.5zm0 9.001a3.5 3.5 0 1 1 3.5-3.5 3.5 3.5 0 0 1-3.5 3.5z"
      />
    </svg>
  </span>
  <span
    class="emotion-2"
  >
    <svg
      fill="currentColor"
      height="24"
      viewBox="0 0 32 32"
      width="24"
    >
      <path
        d="M16.003 24.01c-7.338 0-10.983-5.836-11.89-7.539a.974.974 0 0 1 .003-.927l-.003.005c.907-1.703 4.552-7.539 11.89-7.539s10.983 5.836 11.89 7.539a.974.974 0 0 1-.002.927l.003-.005c-.907 1.703-4.552 7.539-11.89 7.539zm.007-13.5a5.5 5.5 0 1 0 5.5 5.5 5.5 5.5 0 0 0-5.5-5.5zm0 9.001a3.5 3.5 0 1 1 3.5-3.5 3.5 3.5 0 0 1-3.5 3.5z"
      />
    </svg>
  </span>
  <span
    class="emotion-3"
  >
    <svg
      fill="currentColor"
      height="24"
      viewBox="0 0 32 32"
      width="24"
    >
      <path
        d="M16.003 24.01c-7.338 0-10.983-5.836-11.89-7.539a.974.974 0 0 1 .003-.927l-.003.005c.907-1.703 4.552-7.539 11.89-7.539s10.983 5.836 11.89 7.539a.974.974 0 0 1-.002.927l.003-.005c-.907 1.703-4.552 7.539-11.89 7.539zm.007-13.5a5.5 5.5 0 1 0 5.5 5.5 5.5 5.5 0 0 0-5.5-5.5zm0 9.001a3.5 3.5 0 1 1 3.5-3.5 3.5 3.5 0 0 1-3.5 3.5z"
      />
    </svg>
  </span>
  <span
    class="emotion-4"
  >
    <svg
      fill="currentColor"
      height="24"
      viewBox="0 0 32 32"
      width="24"
    >
      <path
        d="M16.003 24.01c-7.338 0-10.983-5.836-11.89-7.539a.974.974 0 0 1 .003-.927l-.003.005c.907-1.703 4.552-7.539 11.89-7.539s10.983 5.836 11.89 7.539a.974.974 0 0 1-.002.927l.003-.005c-.907 1.703-4.552 7.539-11.89 7.539zm.007-13.5a5.5 5.5 0 1 0 5.5 5.5 5.5 5.5 0 0 0-5.5-5.5zm0 9.001a3.5 3.5 0 1 1 3.5-3.5 3.5 3.5 0 0 1-3.5 3.5z"
      />
    </svg>
  </span>
</div>
`;

exports[`IconButton Should match snapshot for with icon 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  padding: 0;
  border: none;
  width: 24px;
  height: 24px;
  color: #ffffff;
  cursor: default;
  pointer-events: none;
  background: #3a44e9;
  background: linear-gradient(122deg,#3a44e9,#7038e3);
}

.emotion-0 > * {
  width: 20px;
  height: 20px;
}

<span
  class="emotion-0"
>
  <svg
    fill="currentColor"
    height="24"
    viewBox="0 0 32 32"
    width="24"
  >
    <path
      d="M16.003 24.01c-7.338 0-10.983-5.836-11.89-7.539a.974.974 0 0 1 .003-.927l-.003.005c.907-1.703 4.552-7.539 11.89-7.539s10.983 5.836 11.89 7.539a.974.974 0 0 1-.002.927l.003-.005c-.907 1.703-4.552 7.539-11.89 7.539zm.007-13.5a5.5 5.5 0 1 0 5.5 5.5 5.5 5.5 0 0 0-5.5-5.5zm0 9.001a3.5 3.5 0 1 1 3.5-3.5 3.5 3.5 0 0 1-3.5 3.5z"
    />
  </svg>
</span>
`;
